enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,761 @@
+* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
+allow access to all the subdomains belonging to the wildcard DNS names.
+
+### Goals
+
+* Support allowing access to DNS names even if the IP addresses associated with them changes
+* Support usage of wildcard DNS names in EgressFirewall rules.
+
+### Non-Goals
+
+* Support additional DNS resolution functionality in the new CoreDNS plugin. The new plugin will only inspect the
+response of DNS resolution by the other existing plugins.
+* Support denying of DNS lookups in the new CoreDNS plugin. The new plugin will not stop the DNS lookup itself and will not
+respond with a `REFUSED`/`NXDOMAIN` response code if a EgressFirewall rule denies access to the specific DNS name being queried for.
+
+## Proposal
+
+This enhancement proposes to introduce a new CoreDNS [external plugin](https://coredns.io/explugins/) (`ocp_dnsnameresolver`) and a new Custom Resource
+(`DNSNameResolver`) to improve the integration of CoreDNS with EgressFirewall. This proposal takes the OVN Interconnect (OVN-IC)
+architecture into consideration where there are (possibly) multiple OVN-K masters and a centralized OVN-K cluster manager. The OVN-K cluster manager will
+create a `DNSNameResolver` CR for each unique DNS name (both regular and wildcard DNS names) used in
+the EgressFirewall rules. The same DNS name may appear in multiple EgressFirewall rules, but only a single `DNSNameResolver` CR object will be used
+for a unique DNS name. This CR will be used to store the DNS name along with the current IP addresses, the corresponding
+TTL, and the last lookup time. The `DNSNameResolver` CR is meant for communication between
+CoreDNS and OVN-K master(s).
+
+The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
+plugins. If the DNS name in the query matches any `DNSNameResolver` CR(s) (regular or wildcard or both), then the
+plugin will update the `.status` of the matching `DNSNameResolver` CR(s) with the DNS name along with the IP addresses,
+the corresponding TTL, and the last lookup time. The OVN-K master(s) will watch the `DNSNameResolver`
+CRs. Whenever the IP addresses are updated for a `DNSNameResolver` CR, the OVN-K master(s) will update the underlying `AddressSet`
+referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).
+
+A new controller (`DNSNameResolver` controller) will keep track of the next lookup time (TTL + last lookup time)
+for each regular DNS name and send a DNS lookup
+query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be only performed
+on the DNS name as it will not return the IP addresses of all the subdomains. The DNS lookup of the wildcard DNS name may fail to return
+any IP address as well. If the lookup for the wildcard DNS name fails, then it will retried using the default TTL (30 minutes). If the lookup
```

Comment:
IIRC we added a retry counter, is it documented somewhere? Perhaps a link here to it would be helpful
```
When a lookup fails consistently over time, the record will be pruned as described in ...
```
WDYT?

Reply:
Not yet resolved?

Reply:
Missed it. Added the link now.